### PR TITLE
feat: [CI-4834]: added acr step

### PR DESCRIPTION
**Pull request by aman-harness**
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35665)

### CODE REVIEW

**Comment by aman-harness** [comment]
trigger ft1

**Comment by aman-harness** [comment]
trigger ft2

**Comment by aman-harness** [comment]
trigger compile

**Comment by aman-harness** [comment]
trigger message

**Comment by aman-harness** [comment]
trigger message

**Comment by aman-harness** [comment]
Automation / Unit tests / Creation of drone plugins are being tracked here https://harness.atlassian.net/browse/CI-5096 https://harness.atlassian.net/browse/CI-5097 https://harness.atlassian.net/browse/CI-5098 and are part of the current sprint

**Comment by aman-harness** [comment]
trigger ti3

File: 332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java.new [hunk 1]
```diff
@@ -0,0 +1 @@
+/*
```
Comment (shubham149) [inline 332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java.new:1]: remove this file

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 2]
```diff
@@ -139,18 +147,19 @@ public static Map<EnvVariableEnum, String> getConnectorSecretEnvMap(CIStepInfoTy
     Map<EnvVariableEnum, String> map = new HashMap<>();
     switch (stepInfoType) {
       case ECR:
-        map.put(EnvVariableEnum.AWS_ACCESS_KEY, PLUGIN_ACCESS_KEY);
```
Comment (shubham149) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:142]: Undo this change since we will add a fix to un-comment L145-146 after new delegate release is complete.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java [hunk 3]
```diff
@@ -462,6 +469,7 @@ private Map<String, String> getAndUpdateGithubAppTokenSecretData(
   private Map<String, String> getAndUpdateConnectorSecretData(
       Map<String, ConnectorDetails> pluginConnectors, CIK8ContainerParams containerParams, String secretName) {
     Map<String, SecretParams> secretData = secretSpecBuilder.decryptConnectorSecretVariables(pluginConnectors);
+    updateSecretForAzure(secretData, pluginConnectors);
```
Comment (shubham149) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:472]: Move this to secretSpecBuilder.decryptConnectorSecretVariables method

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java [hunk 4]
```diff
@@ -470,6 +478,19 @@ private Map<String, String> getAndUpdateConnectorSecretData(
...
 
+  private void updateSecretForAzure(
+      Map<String, SecretParams> secretData, Map<String, ConnectorDetails> pluginConnectors) {
+    for (Map.Entry<String, ConnectorDetails> entry : pluginConnectors.entrySet()) {
+      ConnectorDetails connectorDetails = entry.getValue();
+      if (connectorDetails.getConnectorType() != ConnectorType.AZURE) {
+        return;
+      }
+
+      AzureAcrTokenTaskResponse acrLoginToken = azureAsyncTaskHelper.getAcrLoginToken("testciaman.azurecr.io",
```
Comment (aman-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:489]: checking this hardcoded value.

Comment (shubham149) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:489]: Do this inside the plugin ideally since there can be some deadline for the token & it is possible that token may get expired by the time the step runs.

Comment (shubham149) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:489]: Arguments are hardcoded here.
Also, not setting secret data in this method.

Comment (shubham149) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:489]: Move it to secretSpecBuilder

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java (hunk 3 above)
Comment (aman-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:472]: It can't be moved. Because the logic of talking to the azure sdk is written in 930-delegate tassks and the spec builder is in 933-commons. So this needs to be here in 930. I will talk to CDP team and ask them to move the azure impl in lower module. Then we can move this code to 933-commons

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java (hunk 4 above)
Comment (aman-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java:489]: It can't be moved to secretSpecBuilder as the secretSpecBuilder is in 933-ci-commons and the azure implementation is in 930-commons.

Checking the hardcoded argument.

Comment (aman-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java]: fixed

Comment (aman-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/CIK8InitializeTaskHandler.java]: fixed

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java (hunk 2 above)
Comment (aman-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:142]: done

**Comment by aman-harness** [comment]
trigger ti3

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/execution/CIExecutionConfigService.java [hunk 5]
```diff
@@ -221,6 +227,9 @@ private void applyOverrides(CIExecutionImages defaultConfig, CIExecutionImages o
     if (Strings.isNotBlank(overriddenConfig.getBuildAndPushDockerRegistryTag())) {
       defaultConfig.setBuildAndPushDockerRegistryTag(overriddenConfig.getBuildAndPushDockerRegistryTag());
     }
+    if (Strings.isNotBlank(overriddenConfig.getBuildAndPushACRTag())) {
+      defaultConfig.setBuildAndPushECRTag(overriddenConfig.getBuildAndPushACRTag());
```
Comment (devkimittal) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/execution/CIExecutionConfigService.java:231]: setBuildAndPushACRTag*

Comment (aman-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/execution/CIExecutionConfigService.java]: fixed

**Review by devkimittal — COMMENTED** [review]
> Reviewed 25 of 40 files at r1, 7 of 14 files at r2, 6 of 12 files at r4, 4 of 7 files at r5, 8 of 9 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35665)* status: 3. unresolved discussions. 0 of 1 approvals obtained. Waiting on team(s): sto.

**Review by devkimittal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35665)* status: 2. unresolved discussions. Waiting on team(s): sto.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java [hunk 6]
```diff
@@ -420,6 +427,21 @@ private ConnectorDetails getAwsConnectorDetails(
...
+    List<EncryptedDataDetail> encryptedDataDetails;
+    AzureConnectorDTO azureConnectorDTO = (AzureConnectorDTO) connectorDTO.getConnectorInfo().getConnectorConfig();
+    AzureCredentialDTO credentialDTO = azureConnectorDTO.getCredential();
+    if (credentialDTO.getAzureCredentialType() == MANUAL_CREDENTIALS) {
+      AzureManualDetailsDTO config = (AzureManualDetailsDTO) credentialDTO.getConfig();
+      encryptedDataDetails =
+          secretManagerClientService.getEncryptionDetails(ngAccess, config.getAuthDTO().getCredentials());
+      return connectorDetailsBuilder.encryptedDataDetails(encryptedDataDetails).build();
+    } else {
+      return connectorDetailsBuilder.build();
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java:441]: What about INHERIT_FROM_DELEGATE case ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 7]
```diff
@@ -242,6 +258,67 @@ private static Map<String, String> getGCRStepInfoEnvVariables(
...
+      setOptionalEnvironmentVariable(map, PLUGIN_BUILD_ARGS, mapToStringSlice(buildArgs));
+    }
+
+    Map<String, String> labels =
+        resolveMapParameter("labels", "BuildAndPushACR", identifier, stepInfo.getLabels(), false);
+    if (isNotEmpty(labels)) {
+      setOptionalEnvironmentVariable(map, PLUGIN_CUSTOM_LABELS, mapToStringSlice(labels));
+    }
+
+    if (infraType == Type.K8) {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:301]: can we handle both infra in new method ?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java (hunk 6 above)
Comment (aman-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/ConnectorUtils.java:441]: It's handled in the Drone plugin. In that case, there are no encrypted variables in that case.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/ACRStep.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+package io.harness.ci.states;
+
+import io.harness.beans.steps.stepinfo.ACRStepInfo;
+import io.harness.pms.contracts.steps.StepType;
+
+public class ACRStep extends AbstractStepExecutable {
+  public static final StepType STEP_TYPE = ACRStepInfo.STEP_TYPE;
+
+  // TODO - https://harness.atlassian.net/browse/CI-5084
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/ACRStep.java:16]: what is this ?

Comment (harshjain12) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/ACRStep.java:16]: Is it upload artifact links ?

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/ci/pod/EnvVariableEnum.java [hunk 9]
```diff
@@ -13,6 +13,10 @@ public enum EnvVariableEnum {
   DOCKER_REGISTRY,
   AWS_ACCESS_KEY,
   AWS_SECRET_KEY,
+  AZURE_APP_ID,
```
Comment (harshjain12) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/ci/pod/EnvVariableEnum.java:16]: add them in end because of backward compatibility

File: 970-ng-commons/src/main/java/io/harness/EntityType.java [hunk 10]
```diff
@@ -231,6 +231,9 @@ public enum EntityType {
   @JsonProperty(EntityTypeConstants.BUILD_AND_PUSH_ECR)
   BUILD_AND_PUSH_ECR(ModuleType.CI, EntityTypeConstants.BUILD_AND_PUSH_ECR, IdentifierRef.class,
       EntityYamlRootNames.BUILD_AND_PUSH_ECR),
+  @JsonProperty(EntityTypeConstants.BUILD_AND_PUSH_ACR)
+  BUILD_AND_PUSH_ACR(ModuleType.CI, EntityTypeConstants.BUILD_AND_PUSH_ACR, IdentifierRef.class,
```
Comment (harshjain12) [inline 970-ng-commons/src/main/java/io/harness/EntityType.java:235]: move all newly added enums to end

**Comment by harshjain12** [comment]
@aman-harness Also please add UTs

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java (hunk 7 above)
Comment (aman-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:301]: Done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/ACRStep.java (hunk 8 above)
Comment (aman-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/ACRStep.java:16]: Yes. It's a bit different in the case of azure. I am planning to do it after merging this pr. The ticket is part of sprint and acr epic.

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/ci/pod/EnvVariableEnum.java (hunk 9 above)
Comment (aman-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/ci/pod/EnvVariableEnum.java:16]: done

File: 970-ng-commons/src/main/java/io/harness/EntityType.java (hunk 10 above)
Comment (aman-harness) [reply 970-ng-commons/src/main/java/io/harness/EntityType.java:235]: done

**Comment by aman-harness** [comment]
trigger ti4

**Review by harshjain12 — APPROVED** [review]
:lgtm:

> Reviewed 25 of 40 files at r1, 5 of 14 files at r2, 5 of 12 files at r4, 4 of 7 files at r5, 8 of 9 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35665#-:-N8NytaI2hDxxkfvRmO5:bnfp4nl)* status: 5 file(s) to review. 2. unresolved discussions. Waiting on team(s): sto.

**Review by aman-harness — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35665#-:-N8O8QHRCHGYOt9XHJaI:bnfp4nl)* status: 5 file(s) to review. 3. unresolved discussions. Waiting on team(s): sto. 

___
*[`960-api-services/src/main/java/io/harness/azure/impl/AzureAuthorizationClientImpl.java` line 157 at r7](https://reviewable.io/reviews/harness/harness-core/35665#-N7gn2f41zxLpp_22pRd-r7-157:-N8O8TMkELR46ya1qRPw:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e46dcc1f46fe923feadb3341c89d0d6606c439d5/960-api-services/src/main/java/io/harness/azure/impl/AzureAuthorizationClientImpl.java#L157)):*
<details><summary><i>Previously, harshjain12 (Harsh Jain) wrote…</i></summary><blockquote>

why is it hard coded here ?
can we move it to constants ?
</blockquote></details>

Done.
___
*[`332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java.new` line 1 at r3](https://reviewable.io/reviews/harness/harness-core/35665#-N85vk4Fs9bpy-WD08u5-r3-1:-N8O8V3c0lzUwKOS1eqj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/37e19f4a4dff15697458c4d4abd466d1b42853da/332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java.new#L1)):*
<details><summary><i>Previously, shubham149 (Shubham Agrawal) wrote…</i></summary><blockquote>

remove this file
</blockquote></details>

Done.

File: 332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  private ParameterField<Map<String, String>> labels;
+  @YamlSchemaTypes(value = {string})
+  @ApiModelProperty(dataType = STRING_MAP_CLASSPATH)
+  private ParameterField<Map<String, String>> buildArgs;
+  @YamlSchemaTypes({string}) @ApiModelProperty(dataType = INTEGER_CLASSPATH) private ParameterField<Integer> runAsUser;
+  @YamlSchemaTypes({string}) @ApiModelProperty(dataType = BOOLEAN_CLASSPATH) private ParameterField<Boolean> optimize;
+  @ApiModelProperty(dataType = STRING_CLASSPATH) private ParameterField<String> remoteCacheImage;
+  @YamlSchemaTypes(value = {string})
+  @ApiModelProperty(dataType = STRING_LIST_CLASSPATH)
+  private ParameterField<List<String>> baseImageConnectorRefs;
```
Comment (shubham149) [inline 332-ci-manager/contracts/src/main/java/io/harness/cimanager/stepinfo/ACRStepInfo.java:90]: Is base image connector supported for ACR?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/k8java/CIK8JavaClientHandler.java [hunk 12]
```diff
@@ -89,7 +99,21 @@ public V1Pod createOrReplacePodWithRetries(CoreV1Api coreV1Api, V1Pod pod, Strin
 
   public V1Secret createRegistrySecret(
       CoreV1Api coreV1Api, String namespace, String secretName, ImageDetailsWithConnector imageDetails) {
-    String credentialData = imageSecretBuilder.getJSONEncodedImageCredentials(imageDetails);
+    String credentialData = null;
+    if (imageDetails.getImageConnectorDetails() != null
```
Comment (shubham149) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/citasks/cik8handler/k8java/CIK8JavaClientHandler.java:103]: These should not be a part of this file. It should be present in imageSecretBuilder.

File: 310-ci-manager/ci-manager-config.yml [hunk 13]
```diff
@@ -135,6 +135,11 @@ ciExecutionServiceConfig:
       image: plugins/kaniko-gcr:1.6.1
       entrypoint:
         - /kaniko/kaniko-gcr
+    # I am working on creating a new tag for acr and will update it once that is created.
+    buildAndPushACRConfig:
+      image: bewithaman/kaniko:latest
```
Comment (shubham149) [inline 310-ci-manager/ci-manager-config.yml:140]: Update the repo to use plugins org.
